### PR TITLE
Update Ruby version to 3.1.0

### PR DESCRIPTION
**Pull request by jemmaissroff**
**Before merging:**

- [x] Copy the table printed at the end of the latest benchmark results into the `README.md` and update this PR
- [x] If this change merits an update to `CHANGELOG.md`, add an entry following Keep a Changelog [guidelines](https://keepachangelog.com/en/1.0.0/) with [semantic versioning](https://semver.org/)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,7 +7,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
-- Nothing yet!
+- Bump codebase to Ruby 3.1.0
```
Comment (JacobEvelyn) [inline CHANGELOG.md:10]: Maybe something like "Update official test coverage to support Ruby 3.1"? My thinking is that the code didn't change.

File: spec/thread_safety_spec.rb [hunk 2]
```diff
@@ -31,13 +31,15 @@ def current_thread_id(#{args_str})
     # which caused accidental `nil` values to be returned under contended calls
     # to unmemoized zero-args methods.
     #   * See https://github.com/panorama-ed/memo_wise/pull/224
-    context "when checking condition: are nil values accidentally returned?" do
-      let(:condition_to_check) do
-        ->(values) { values.any?(&:nil?) }
-      end
+    unless RUBY_PLATFORM == "java"
```
Comment (JacobEvelyn) [inline spec/thread_safety_spec.rb:34]: Do we not want to run these on MRI? Can you add some context in a comment?

File: README.md [hunk 3]
```diff
@@ -115,19 +115,19 @@ For more usage details, see our detailed [documentation](#documentation).
...
-|`(a)`|1.99x|10.19x|
-|`(a, b)`|0.39x|1.94x|
-|`(a:)`|1.80x|19.36x|
-|`(a:, b:)`|0.43x|4.26x|
-|`(a, b:)`|0.39x|3.97x|
-|`(a, *args)`|0.83x|1.85x|
-|`(a:, **kwargs)`|0.77x|2.94x|
-|`(a, *args, b:, **kwargs)`|0.59x|1.57x|
+|`()` (none)|1.09x|11.80x|
+|`(a)`|1.64x|9.74x|
```
Comment (JacobEvelyn) [inline README.md:123]: Interesting; some of these seem like big changes 🤷‍♀️

File: spec/thread_safety_spec.rb (hunk 2 above)
Comment (jemmaissroff) [reply spec/thread_safety_spec.rb:33]: Added some context, and #264. Let me know how it looks now

**Review by JacobEvelyn — APPROVED** [review]

**Pull request merged**